### ISSUE TITLE
Bad file descriptor problem still exists as of today

### ISSUE THREAD

**Issue by crankedguy** [benoitc/gunicorn#3013]
Hello,

why was this
https://github.com/benoitc/gunicorn/issues/1877
closed and further comments after closure completely ignored?

I must admit the crosslinking between the project issues and the conversation is extremely hard to follow, for me honestly not at all anymore now what is done, reverted, has to be done, etc.
https://github.com/encode/uvicorn/pull/1710

Can someone please shed some light onto if and how this will be fixed, because this is still an issue as of today with gunicorn 20.1.- and uvicorn / uvicornworkers 0.22, and is also very easy to reproduce every single time the server is shutdown (e.g. with Ctrl-C)

Maybe 
@Kludex ??

Answer highly appreciated, thanks

**Comment by Kludex** [comment]
To me, what I read here is: "I don't want to read all the issues, can you summarize for me?".

As I mentioned on https://github.com/benoitc/gunicorn/issues/1877#issuecomment-1296752389, I believe the problem is on `gunicorn`. On that message, I also mention that this error will appear more frequently, since there was a bigger issue (the shutdown event was not being triggered, which was solved by https://github.com/encode/uvicorn/pull/1710).

I've also asked about the `SIGQUIT` behavior almost 2 years ago on #2604, and on https://github.com/benoitc/gunicorn/issues/2604#issuecomment-874705562 @benoitc said he'll look for it, and I was ignored, and the issue was further closed with this comment: https://github.com/benoitc/gunicorn/issues/2604#issuecomment-1278793353.

So... I've already offered, and dedicated too much of my time. Also, I find it very rude of you for pinging me here. 👍

**Comment by crankedguy** [comment]
Actually I did read it, but at some point I got simply lost. There are countless threads involved, changes implemented, reverted.. I also saw that this was more or less ignored on this side sadly. As fas as I understood the SIGQUIT handler was implemented in uvicorn worker but the problem persists...

On being rude : Although I can't really see a rudeness in a simple question, it was certainly not the intention if you consider it like that. So you can take my sorry for it. I won't bother any further...

**Comment by Kludex** [comment]
> Although I can't really see a rudeness in a simple question

🤷‍♂️ 

> sorry for it.

Apologize accepted.

> As fas as I understood the SIGQUIT handler was implemented in uvicorn worker but the problem persists...

The problem that was solved is the shutdown event not running when using uvicorn, which was indeed solved. The compromise was having the message "Bad file descriptor" more often on the logs.

The problem that needs to be solved is mentioned here: https://github.com/benoitc/gunicorn/issues/1877#issuecomment-1296752389.

**Comment by crankedguy** [comment]
> The problem that needs to be solved is mentioned here: [#1877 (comment)](https://github.com/benoitc/gunicorn/issues/1877#issuecomment-1296752389).

Yes I also saw that one. That comment came after that particular issue was just closed, and there was no activity on it except from users still having the problem :pray:

Sometimes you only get the already documented 
`[2023-06-21 16:13:39 +0200] [1238732] [INFO] Error while closing socket [Errno 9] Bad file descriptor`

and when you get - also already documented - 
`[2023-06-21 16:13:39 +0200] [1180582] [WARNING] Worker with pid 1238730 was terminated due to signal 15`
you have these trace remainings in job output
```
Jun 21 16:13:39 domain.com gunicorn[1180582]:   File "/usr/lib64/python3.11/logging/__init__.py", line 1094, in flush
Jun 21 16:13:39 domain.com gunicorn[1180582]:     self.stream.flush()
Jun 21 16:13:39 domain.com gunicorn[1180582]:   File "/home/usr/.venvs/xxx_api/lib64/python3.11/site-packages/gunicorn/arbiter.py", line 242, in handle_chld
Jun 21 16:13:39 domain.com gunicorn[1180582]:     self.reap_workers()
Jun 21 16:13:39 domain.com gunicorn[1180582]:   File "/home/usr/.venvs/xxx_api/lib64/python3.11/site-packages/gunicorn/arbiter.py", line 530, in reap_workers
Jun 21 16:13:39 domain.com gunicorn[1180582]:     self.log.warning(
Jun 21 16:13:39 domain.com gunicorn[1180582]:   File "/home/usr/.venvs/xxx_api/lib64/python3.11/site-packages/gunicorn/glogging.py", line 261, in warning
Jun 21 16:13:39 domain.com gunicorn[1180582]:     self.error_log.warning(msg, *args, **kwargs)
Jun 21 16:13:39 domain.com gunicorn[1180582]: Message: 'Worker with pid %s was terminated due to signal %s'
Jun 21 16:13:39 domain.com gunicorn[1180582]: Arguments: (1238732, 15)
```

**Comment by benoitc** [comment]
@Kludex you weren't ignored. I offered a response to your issue that didn't match your expectation at this time. This happen sometimes, but don't say it was ignored... 

@crankedguy that's the good approach. All issues closed are closed and if a new problem arise or if an issue still exist a new ticket should be created. This is described in the contributing doc.

About this issue. I didn't have yet re-tested latest master with uvicorn. This is planned this we. I have been side tracked with some coming changes in the http parser and some other changes for the new release. The behavior is not reproduced with others workers, I will try to understand what is the design difference there that causes this issue.

**Comment by crankedguy** [comment]
That would be great because this is kinda a problem not only with command line invoke instances but also if you habdle them over systemd which leads to unclean unit shutdowns

**Comment by MichaelOdumosu57** [comment]
What worked for me if watchdog and socketio was being used in the application (in this case flask) the error would show up as socket is also a file in way. Hopefully this is useful and it helps

**Comment by tilgovi** [comment]
In [this comment](https://github.com/benoitc/gunicorn/issues/1877#issuecomment-739526302), @gg349 very helpfully tracked down [the uvicorn commit](https://github.com/encode/uvicorn/commit/c4900d19e1100a7b1a93a99f3d3ec6b717ffea41) that re-ordered calls to close the asyncio servers and close the sockets.

In the asyncio documentation for [loop.create_server](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_server), it says:

> Note: The sock argument transfers ownership of the socket to the server created. To close the socket, call the server’s [close()](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.Server.close) method.

That means that `uvicorn` should not call `socket.close()` because it has given ownership of the socket to the server. Before the referenced commit, uvicorn was closing the socket first and uvloop was [tolerating EBADF](https://github.com/MagicStack/uvloop/blob/master/uvloop/handles/handle.pyx#L283-L289) for its own internal reasons. After the commit, uvicorn is raising an EBADF because the server has already closed the socket.

There is nothing gunicorn can do for this issue. The error happens in the worker, where uvicorn is closing the socket after it was already closed.

I suggest we close the issue here if this diagnosis makes sense to @crankedguy and @Kludex.

**Comment by tilgovi** [comment]
Note that this only happens with the uvloop dependency.

Python sockets can actually be closed multiple times without issue. Python tracks when a socket has been closed and only issues a close to the OS once. However, if you close a file descriptor out from underneath the Python socket, then closing it from Python will cause an EBADF error:

```python
>>> s = socket.socket(socket.SOCK_STREAM)
>>> socket.close(s.fileno())
>>> s.close()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.11/socket.py", line 503, in close
    self._real_close()
  File "/usr/lib/python3.11/socket.py", line 497, in _real_close
    _ss.close(self)
OSError: [Errno 9] Bad file descriptor
```

Gunicorn actually passes a `gunicorn.sock.TCPSocket` object, a wrapper around the Python `socket` object. Gunicorn knows not to close the underlying socket twice, even though doing so should be fine.

So, it doesn't normally cause problems that uvicorn closes both the server and the socket. However, things get weird with uvloop and gunicorn.

When the server gets closed, [uvloop calls `uv_close()`](https://github.com/MagicStack/uvloop/blob/master/uvloop/handles/handle.pyx#L212) in the underlying libuv. The [uv_close() documentation](https://docs.libuv.org/en/v1.x/handle.html#c.uv_close) says:

> Handles that wrap file descriptors are closed immediately but close_cb will still be deferred to the next iteration of the event loop. It gives you a chance to free up any resources associated with the handle.

In other words, socket handles in uvloop expect that the base handle has already closed the fd and want to simply call `detach()` on the Python socket. Since the gunicorn socket isn't an instance of the built-in `socket` type, [uvloop calls `close()` instead of `detach()`](https://github.com/MagicStack/uvloop/blob/master/uvloop/handles/handle.pyx#L272-L281). The `gunicorn.sock.TCPSocket` wrapper calls `close()` on the real socket even though its file descriptor is already closed by libuv, raising the EBADF.

Gunicorn could add a defensive `try/except` in its socket wrapper, but I don't think it should. Regular Python sockets don't do that. Instead, uvicorn should rely on servers to close their sockets, as documented in the asyncio module.

**Comment by tilgovi** [comment]
If uvicorn cannot track which sockets it has and has not passed to servers, it may want to have its own defensive `try/except`. Better might be to pass `s.sock for s in sockets` from the `UvicornWorker`, so that uvicorn is working with the underlying sockets.

If we consider any change in gunicorn itself, I'd consider trying to remove the socket wrapper classes, or making them actually inherit from `socket`.

**Comment by tilgovi** [comment]
I've opened a [draft PR](https://github.com/benoitc/gunicorn/pull/3127) for getting rid of the socket wrapper classes and verified that this fixes this issue with uvloop.

**Comment by jsprieto10** [comment]
thanks @tilgovi Waiting for you PR to be ready

**Comment by cometaj2** [comment]
I seem to have a similar but slightly differentiated version of this same underlying socket/file descriptor error? It runs all the way thru to urllib3's pool cleanup. This happens with gunicorn on both sync and gevent workers but otherwise works flawlessly with waitress. The failure under gunicorn stalls the worker or thread and hampers execution continuation in a Flask PWA.

$ gunicorn --workers=1 --threads=100 -b 127.0.0.1:8000 "hello:app" (stalls; gunicorn==23.0.0)
$ waitress-serve --listen=127.0.0.1:8000 --threads=100 hello:app (succeeds; waitress==3.0.2)

I can circumvent the issue on gunicorn by having the requests HTTP client force close the connection on every request (connection: close header). The error still occurs with urllib3 pool cleanup and there's a slight delay on what I assume to be some internal shenanigans as the connection fails and times out but the PWA application is able to continue working.

Exception ignored in: <finalize object at 0x106c9bd20; dead>
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.11/3.11.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/weakref.py", line 590, in __call__
    return info.func(*info.args, **(info.kwargs or {}))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jeff/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 1138, in _close_pool_connections
    conn.close()
  File "/opt/homebrew/Cellar/python@3.11/3.11.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/http/client.py", line 981, in close
    sock.close()   # close it manually... there may be other refs
    ^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/socket.py", line 503, in close
    self._real_close()
  File "/opt/homebrew/Cellar/python@3.11/3.11.11/Frameworks/Python.framework/Versions/3.11/lib/python3.11/socket.py", line 497, in _real_close
    _ss.close(self)
OSError: [Errno 9] Bad file descriptor
